### PR TITLE
New logic: Automatic Passenger Deletion

### PR DESCRIPTION
**Pull request by FS-21**
New tags applied in a transport for consuming the passengers after some frames.
PassengerDeletion.Countdown=0           ; integer. Game frames multiplied by the unit size
PassengerDeletion.Soylent=no            ; boolean
PassengerDeletion.SoylentFriendlies=no  ; boolean
PassengerDeletion.Report=               ; sound

Can be combined with Ares abductor logic for a better Chrono Prison & optionally this can refund some money like a Grinder.

### CODE REVIEW

**Comment by FS-21** [comment]
Because if the logic could be used only for the grinding feature I added a new tag:
PassengerDeletion.SizeDependence=yes
By default the Unit's size is used for calculating the unit's countdown for the deletion. If set "No" then only the PassengerDeletion.Countdown value is used. In that case use PassengerDeletion.Countdown=1 for instant passenger deletion.

File: docs/New-or-Enhanced-Logics.md [hunk 1]
```diff
@@ -223,6 +223,21 @@ NotHuman.RandomDeathSequence=yes  ; boolean
...
+ - Good combination with Ares Abductor logic.
+
+In `rulesmd.ini`:
+```ini
+[SOMETECHNO]                            ; TechnoType
+PassengerDeletion.Countdown=0           ; integer, game frames
+PassengerDeletion.Soylent=no            ; boolean
+PassengerDeletion.SoylentFriendlies=no  ; boolean
+PassengerDeletion.SizeDependence=yes    ; boolean. Final countdown is done multiplying the Unit size
+PassengerDeletion.Report=               ; sound
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:238]: See tag name suggestions.
```suggestion
PassengerDeletion.Rate=                 ; integer, game frames
PassengerDeletion.Rate.SizeMultiply=yes ; boolean, whether to multiply frames amount by size
PassengerDeletion.Soylent=no            ; boolean
PassengerDeletion.SoylentFriendlies=no  ; boolean
PassengerDeletion.ReportSound=          ; sound
```

File: src/Ext/Techno/Body.h [hunk 2]
```diff
@@ -22,12 +22,14 @@ class TechnoExt
 		std::unique_ptr<ShieldClass> Shield;
 		ValueableVector<std::unique_ptr<LaserTrailClass>> LaserTrails;
 		Valueable<bool> ReceiveDamage;
+		int PassengerDeletion_Countdown;
```
Comment (Metadorius) [inline src/Ext/Techno/Body.h:25]: Not sure why this is not a timer.

File: src/Ext/TechnoType/Body.h [hunk 3]
```diff
@@ -39,6 +39,11 @@ class TechnoTypeExt
 		Valueable<int> CameoPriority;
 		Valueable<bool> NoManualMove;
 		Nullable<int> InitialStrength;
+		Valueable<bool> PassengerDeletion_Refund;
+		Valueable<bool> PassengerDeletion_RefundFriendlies;
+		Valueable<int> PassengerDeletion_Countdown;
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.h:44]: Maybe `Nullable` will be better here?

**Review by Metadorius — CHANGES_REQUESTED** [review]

**Review by Metadorius — CHANGES_REQUESTED** [review]
More style issues that I missed from the last time. Also what about other concerns that I have raised during the last review?

File: src/Ext/Techno/Body.cpp [hunk 4]
```diff
@@ -247,6 +247,103 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
+					// Check if there is money refund
+					if (pTypeData->PassengerDeletion_Refund)
+					{
+						int soylent = 0;
+
+						// Refund money to the Attacker
+						if (pTypePassenger && pTypePassenger->Soylent > 0)
+						{
+							soylent = pTypePassenger->Soylent;
+						}
```
[the remark this answers is not in the document]
Comment (FS-21) [reply src/Ext/Techno/Body.cpp]: changed

File: src/Ext/Techno/Body.cpp [hunk 5]
```diff
@@ -247,6 +247,103 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
+	auto pTypeData = TechnoTypeExt::ExtMap.Find(pThis->GetTechnoType());
+
+	if (pTypeData && pTypeData->PassengerDeletion_Countdown > 0)
+	{
+		auto pData = TechnoExt::ExtMap.Find(pThis);
+
+		if (pThis->Passengers.NumPassengers > 0)
+		{
+			if (pData->PassengerDeletion_Countdown > 0)
+				pData->PassengerDeletion_Countdown--;
```
[the remark this answers is not in the document]
Comment (FS-21) [reply src/Ext/Techno/Body.cpp]: changed

File: src/Ext/Techno/Body.cpp [hunk 6]
```diff
@@ -247,6 +247,103 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
 
+void TechnoExt::EatPassengers(TechnoClass* pThis)
+{
+	if (pThis->TemporalTargetingMe 
+		|| pThis->BeingWarpedOut 
+		|| pThis->IsUnderEMP() 
+		|| !pThis->IsAlive 
+		|| pThis->Health <= 0 
+		|| pThis->InLimbo)
+		return;
```
[the remark this answers is not in the document]
Comment (FS-21) [reply src/Ext/Techno/Body.cpp]: changed

File: src/Ext/Techno/Body.cpp [hunk 7]
```diff
@@ -247,6 +247,103 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
 	return FLH;
 }
 
+void TechnoExt::EatPassengers(TechnoClass* pThis)
```
[the remark this answers is not in the document]
Comment (FS-21) [reply src/Ext/Techno/Body.cpp:250]: Solved?

**Review by FS-21 — COMMENTED** [review]
I think that it's done, maybe.

File: src/Ext/TechnoType/Body.cpp [hunk 8]
```diff
@@ -119,9 +119,9 @@ void TechnoTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
 
 	this->PassengerDeletion_Refund.Read(exINI, pSection, "PassengerDeletion.Soylent");
 	this->PassengerDeletion_RefundFriendlies.Read(exINI, pSection, "PassengerDeletion.SoylentFriendlies");
-	this->PassengerDeletion_Report.Read(exINI, pSection, "PassengerDeletion.Report");
-	this->PassengerDeletion_SizeDependence.Read(exINI, pSection, "PassengerDeletion.SizeDependence");
-	this->PassengerDeletion_Countdown.Read(exINI, pSection, "PassengerDeletion.Countdown");
+	this->PassengerDeletion_Report.Read(exINI, pSection, "PassengerDeletion.ReportSound");
+	this->PassengerDeletion_SizeDependence.Read(exINI, pSection, "PassengerDeletion.Rate.SizeMultiply");
+	this->PassengerDeletion_Countdown.Read(exINI, pSection, "PassengerDeletion.Rate");
```
[the remark this answers is not in the document]
Comment (FS-21) [reply src/Ext/TechnoType/Body.cpp]: renamed.

File: src/Ext/TechnoType/Body.h (hunk 3 above)
Comment (FS-21) [reply src/Ext/TechnoType/Body.h]: PassengerDeletion_Rate should have always a value for if-else comparations :-/

**Review by FS-21 — COMMENTED** [review]
reviewed without mistakes? :-/

**Review by Metadorius — APPROVED** [review]

**Comment by Metadorius** [comment]
Good to go, resolve conflicts and you can merge this one.

**Comment by FS-21** [comment]
Fixed merge conflicts.

**Pull request merged**